core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -507,86 +518,126 @@ trait EntityComponent {
-      val startingEntityIdsAction = filter(rec => rec.workspaceId === workspaceContext.workspaceId && rec.entityType === entityType && rec.name.inSetBind(entityNames)).map(_.id)
+    def getEntitySubtrees(workspaceContext: SlickWorkspaceContext, entityType: String, entityNames: Seq[String]): ReadAction[Seq[Seq[AttributeEntityReference]]] = {
+      val startingEntityIdsAction = filter(rec => rec.workspaceId === workspaceContext.workspaceId && rec.entityType === entityType && rec.name.inSetBind(entityNames))
 
       startingEntityIdsAction.result.flatMap { startingEntityIds =>
-        val idSet = startingEntityIds.toSet
-        recursiveGetEntityReferenceIds(Down, idSet, idSet)
-      } flatMap { ids =>
-        unmarshalEntities(EntityAndAttributesRawSqlQuery.actionForIds(ids))
+        val idSet = startingEntityIds.map(rec => Seq(AttributeEntityReference(rec.entityType, rec.name)) -> rec.id).toMap
+        recursiveGetEntityReferenceIds(Down, startingEntityIds.map(_.id).toSet, startingEntityIds.map(rec => rec.id -> AttributeEntityReference(rec.entityType, rec.name)).toMap, idSet)
       }
     }
 
-    def copyEntities(sourceWorkspaceContext: SlickWorkspaceContext, destWorkspaceContext: SlickWorkspaceContext, entityType: String, entityNames: Seq[String]): ReadWriteAction[TraversableOnce[Entity]] = {
-      getEntitySubtrees(sourceWorkspaceContext, entityType, entityNames).flatMap { entities =>
-        getCopyConflicts(destWorkspaceContext, entities).flatMap { conflicts =>
-          if (conflicts.isEmpty) {
-            cloneEntities(destWorkspaceContext, entities).map(_ => Seq.empty[Entity])
-          } else {
-            DBIO.successful(conflicts)
-          }
+    def copyEntities(sourceWorkspaceContext: SlickWorkspaceContext, destWorkspaceContext: SlickWorkspaceContext, entityType: String, entityNames: Seq[String], linkExistingEntities: Boolean): ReadWriteAction[EntityCopyResponse] = {
+      def getHardConflicts(workspaceId: UUID, entityRefs: Seq[AttributeEntityReference]) = {
+        val batchActions = createBatches(entityRefs.toSet).map(batch => lookupEntitiesByNames(workspaceId, batch))
+        DBIO.sequence(batchActions).map(_.flatten.toSeq).map { recs =>
+          recs.map(rec => AttributeEntityReference(rec.entityType, rec.name))
         }
       }
-    }
 
-    def getCopyConflicts(destWorkspaceContext: SlickWorkspaceContext, entitiesToCopy: TraversableOnce[Entity]): ReadAction[TraversableOnce[Entity]] = {
-      val entityQueries = entitiesToCopy.map { entity =>
-        findEntityByName(destWorkspaceContext.workspaceId, entity.entityType, entity.name).result.map {
-          case Seq() => None
-          case _ => Option(entity)
+      def getSoftConflicts(paths: Seq[Seq[AttributeEntityReference]]) = {
+        getCopyConflicts(destWorkspaceContext, paths.map(_.last)).map { conflicts =>
```

Comment:
if we have the entity tree:

```
sample_set
   ^- sampleA
      ^- participant1
   ^- sampleB
      ^- participant2
```

we should have something like the following for the `paths`:

```
Seq(
Seq(sample_set), 
Seq(sample_set, sampleA), 
Seq(sample_set, sampleA, participant1), 
Seq(sample_set, sampleB), 
Seq(sample_set, sampleB, participant2)
)
```

So we'll pull off sample_set, sampleA, participant1, sampleB, and participant2. I suppose this is equivalent to a flattened and de-duped `paths` but this is just how my brain spit it out as I was imagining the whole thing... but that doesn't mean it's the most clear way for me to write it :)

Of the top of my head I can't think of a case where these two ways of writing could be different